### ISSUE TITLE
Unexpected Results in Local, FedAvg, and Oracle Settings.

### ISSUE THREAD

**Issue by Hjh233** [rui-ye/OpenFedLLM#43]
Thanks for your excellent work! This repository is easy to follow and well-documented. However, I am experiencing some unexpected results across three settings: local, fedavg, and oracle (where one client owns all the data).
### Local
```
batch_size=16
dataset_name="vicgalle/alpaca-gpt4"
dataset_sample=100000
fed_alg="local0"
fed_strategy="local"
gradient_accumulation_steps=1
gpu=2
load_in_8bit=True
# local_data_dir=""       # you may uncomment this line if your data is stored locally and include it in the python command
lora_alpha=64   # twice of lora_r
lora_r=32
lr=5e-5
max_steps=10
model_name_or_path="NousResearch/Llama-2-7b-hf" # refer to https://huggingface.co/NousResearch/Llama-2-7b-hf
num_clients=4
num_train_epochs=5
num_rounds_list=(1 10 100) # aggregation rounds
output_dir=output
sample_clients=4
seed=0
seq_length=512
use_peft=True
```
### fedavg
```
batch_size=16
dataset_name="vicgalle/alpaca-gpt4"
dataset_sample=100000
fed_alg="fedavg"
fed_strategy="fedlora"
gradient_accumulation_steps=1
gpu=0
load_in_8bit=True
# local_data_dir=""       # you may uncomment this line if your data is stored locally and include it in the python command
lora_alpha=64   # twice of lora_r
lora_r=32
lr=5e-5
max_steps=10
model_name_or_path="NousResearch/Llama-2-7b-hf" # refer to https://huggingface.co/NousResearch/Llama-2-7b-hf
num_clients=4
num_train_epochs=5
num_rounds_list=(1 10 100) # aggregation rounds
output_dir=output
sample_clients=4
seed=0
seq_length=512
use_peft=True
```
### oracle
```
batch_size=16
dataset_name="vicgalle/alpaca-gpt4"
dataset_sample=100000
fed_alg="local0"
fed_strategy="oracle"
gradient_accumulation_steps=1
gpu=3
load_in_8bit=True
# local_data_dir=""       # you may uncomment this line if your data is stored locally and include it in the python command
lora_alpha=64   # twice of lora_r
lora_r=32
lr=5e-5
max_steps=10
model_name_or_path="NousResearch/Llama-2-7b-hf" # refer to https://huggingface.co/NousResearch/Llama-2-7b-hf
num_clients=1
num_train_epochs=5
num_rounds_list=(1 10 100) # aggregation rounds
output_dir=output
sample_clients=1
seed=0
seq_length=512
use_peft=True
```
The difference from the original setup is that I use all samples per client in every round by setting:
```
num2sample = len(dataset)
```
in `federated_learning/split_dataset.py/get_dataset_this_round`.
Additionally:
1. I set num_train_epochs to 5 instead of 3.
2. I ran 1/5/10/50/100 rounds instead of 200 (as in the original paper).
3. I consider a cross-silo setting where num_clients=sample_clients=4 in FedAvg.
4. I set the seed to 0.
## Training
![Xnip2025-01-13_14-08-44](https://github.com/user-attachments/assets/855308f7-e207-43c5-a775-be95c7bdb74a)
The loss for oracle setting seems to fluctuate.
## Evaluation
After training, the trained LoRA model is saved at `"path/checkpoint_1"` for example. To merge the model, I run
```
python merge_lora.py --base_model_path "NousResearch/Llama-2-7b-hf" --lora_path "path/checkpoint_1"
```
and I have the full fine-tuned model at `"path/full_1"`.
I followed the evaluation process from [instruct-eval](https://github.com/declare-lab/instruct-eval), running:
```
python main.py mmlu --model_name llama --model_path path/full_1
```
I report MMLU, BBH, DROP, Humaneval, and CRASS scores as per the original paper. However, the results are quite odd.
![Xnip2025-01-13_14-16-00](https://github.com/user-attachments/assets/8512bfb2-2bbe-4141-8855-0c510cf06231)
According to the table in the original paper, I would expect large gaps between local and fedavg/oracle in MMLU, DROP, Humaneval, and CRASS. However, the results are quite close. Additionally, I see identical scores in CRASS:
* 40.99 for local_5_rounds, local_10_rounds, fedavg_5_rounds; 
* 42.24 for local_100_rounds, oracle_10_rounds, oracle_50_rounds. 
I wonder if you have experienced such results before.
## Summary
I expect to see local << fedavg ~ oracle in MMLU, DROP, Humaneval, CRASS. 
However, the results are counter-intuitive, and I’m unsure if this is due to incorrect hyperparameters or some other issue. I would greatly appreciate any insights or advice on this!
Thank you in advance for your help!

**Comment by rui-ye** [comment]
Hi,

MMLU, BBH, DROP, HumanEval, and CRASS are benchmarks that can be used to evaluate pre-trained models, In this case, we report these results just to ensure that the training process would not compromise the capability learned by pre-trained models. It is likely to happen that Local, FedAvg, and oracle perform similarly in these benchmarks.

For this task, we are more interested at the open-ended benchmarks. Also, please notice that in FedLLM, the relationships among local, FedAvg, and Oracle depend on `the number of rounds` and `data amount per client`.

**Comment by Hjh233** [comment]
Thank you so much for your response!
I checked the MMLU, BBH, DROP, HumanEval, and CRASS scores of the pretrained LLaMA-7B model, which serves as the basis for our fine-tuning. It turns out that Local, FedAvg, and oracle outperform the pretrained model in almost all metrics by a relatively large margin. This suggests that the previously observed unusual results might be due to that the data amount for local is already sufficient.
I have two more questions though,
1. **Hyper-parameter Settings**:
I wonder if you could provide a hyper-parameter setting where the performance is local << fedavg as shown in Table 4 of the original paper, including dataset_sample, lora_alpha, lora_r, lr, num_clients, num_train_epochs, num_rounds_list, and sample_clients.
2. **Template Setting for Open-ended Benchmark**:
I am a little bit confused about the template setting w.r.t. model answer generation in open-ended benchmark.
I fine-tuned a LLaMA-7B model with 8-bit quantization on Alpaca-GPT4 dataset and would like to evaluate on open-ended benchmark. 
For MT-bench Step 1. Generate model answers to MT-bench questions:
which template should I choose in [utils/conversation.py](https://github.com/rui-ye/OpenFedLLM/blob/main/utils/conversation.py)?
```
python gen_model_answer_mt.py --base_model_path [MODEL_PATH] --template [TEMPLATE] --lora_path [LORA_PATH]
```
For Vicuna Step 1. Generate model answers
should I use the following command and choose `vicuna` as the template?
```
python gen_model_answer.py --base_model_path [MODEL_PATH] --template vicuna --bench_name [BENCH_NAME] --use_vllm
```
I really appreciate your help, and thank you for the excellent work!
